Modules/_functoolsmodule.c

```diff
@@ -328,8 +505,39 @@ partial_call(partialobject *pto, PyObject *args, PyObject *kwargs)
     }
 
     /* Merge positional arguments */
-    /* Note: tupleconcat() is optimized for empty tuples */
-    PyObject *args2 = PySequence_Concat(pto->args, args);
+    PyObject *args2;
```

Comment:
I would first check whether you can merge the positional arguments before merging the keyword arguments.

Reply:
In C version that check is at the top, before any of the mergings. Line 477.